helix-rest/src/main/java/org/apache/helix/rest/metadatastore/TrieRoutingData.java

```diff
@@ -124,8 +128,91 @@ private TrieNode findTrieNode(String path, boolean findLeafAlongPath)
     return curNode;
   }
 
-  // TODO: THE CLASS WILL BE CHANGED TO PRIVATE ONCE THE CONSTRUCTOR IS CREATED.
-  static class TrieNode {
+  /**
+   * Checks for the edge case when the only sharding key in provided routing data is the delimiter
+   * or an empty string. When this is the case, the trie is valid and contains only one node, which
+   * is the root node, and the root node is a leaf node with a realm address associated with it.
+   * @param routingData - a mapping from "sharding keys" to "realm addresses" to be parsed into a
+   *          trie
+   * @return whether the edge case is true
+   */
+  private boolean isRootShardingKey(Map<String, List<String>> routingData) {
+    if (routingData.values().size() == 1) {
+      for (List<String> shardingKeys : routingData.values()) {
+        return shardingKeys.size() == 1
+            && (shardingKeys.get(0).equals(DELIMITER) || shardingKeys.get(0).equals(""));
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Constructs a trie based on the provided routing data. It loops through all sharding keys and
+   * construct the trie in a top down manner.
+   * @param routingData- a mapping from "sharding keys" to "realm addresses" to be parsed into a
+   *          * trie
+   * @throws InvalidRoutingDataException - when there is an empty sharding key (edge case that
+   *           always renders the routing data invalid); when there is a sharding key which already
+   *           contains a sharding key (invalid); when there is a sharding key that is a part of
+   *           another sharding key (invalid)
+   */
+  private void constructTrie(Map<String, List<String>> routingData)
+      throws InvalidRoutingDataException {
+    for (Map.Entry<String, List<String>> entry : routingData.entrySet()) {
+      for (String shardingKey : entry.getValue()) {
+        // Add a leading delimiter if there isn't any
+        if (!shardingKey.substring(0, 1).equals(DELIMITER)) {
+          shardingKey = DELIMITER + shardingKey;
+        }
+
+        // Root can only be a sharding key if it's the only sharding key. Since this method is
+        // running, the special case has already been checked, therefore it's definitely invalid
+        if (shardingKey.equals(DELIMITER)) {
```

Comment:
You can check this first `isEmpty(), throw exception` before you add DELIMITER to the shardingKey. Could've saved one string object for this loop.

Reply:
That is untrue, because `shardingKey` could be either empty or "/". Adding delimiter first makes the logic easy to understand - after adding the delimiter, all logic applies to sharding keys with leading delimiters.

Reply:
Behaviour has changed. No leading slash is an exception now. Closing conversation.